crates/assists/src/utils.rs

```diff
@@ -338,3 +338,33 @@ fn has_fn(imp: &ast::Impl, rhs_name: &str) -> bool {
 
     false
 }
+
+/// Find the start of the `impl` block for the given `Impl`, or create one if
+/// none exists.
+///
+/// This method is intended to be used in conjunction with
+/// `utils::find_struct_impl`, and `AssistBuilder::insert`.
+//
+// FIXME: include a way to append methods to the end of the `impl` block.
+pub(crate) fn find_or_create_impl_block<F>(
+    impl_def: Option<ast::Impl>,
```

Comment:
```suggestion
    impl_def: ast::Impl,
```

this'll also allow to keep the closure out of the API

cc https://github.com/rust-analyzer/rust-analyzer/blob/master/docs/dev/style.md#function-preconditions

Reply:
ah yes, fair enough! -- I think you just pinpointed some of what felt bad about this impl. Thanks, will change!